utils/src/Vulkan/Utils/Reflect.hs

```diff
@@ -0,0 +1,457 @@
+import Data.Aeson.Types (prependFailure, unexpected)
+import Data.Text (Text)
+import qualified Data.Text as T
+import qualified Data.Text.IO as T
+import Data.Vector (Vector)
+import qualified Data.Vector as V
+import System.FilePath ((</>))
+import System.Process.Typed (proc, readProcess)
+import System.IO.Temp (withSystemTempDirectory)
+import Control.Monad.IO.Class (MonadIO, liftIO)
+
+import Data.List (sortOn, groupBy)
+import Data.Map (Map)
+import qualified Data.Map as M
+import Data.Function (on)
+import Data.Maybe (fromMaybe, catMaybes)
+import Data.Word (Word32)
+import Control.Applicative (liftA2)
+import Control.Monad (join)
+import Control.Exception (Exception (..), throw)
+
+data EntryPoint = EntryPoint
+  { name :: Text
+  , mode :: ShaderStage
+  } deriving (Show, Generic, FromJSON, ToJSON)
+
+data Input = Input
+  { type' :: VertexAttributeType
+  , name :: Text
+  , location :: Int
+  , array :: Maybe (Vector Int)
+  } deriving (Show, Generic)
+
+instance FromJSON Input where
+  parseJSON = withObject "inputs" $ \v -> Input
+    <$> v .: "type"
+    <*> v .: "name"
+    <*> v .: "location"
+    <*> v .:? "array"
+
+instance ToJSON Input where
+  toJSON (Input type' name location array) = object
+    [ "type" .= type'
+    , "name" .= name
+    , "location" .= location
+    , "array" .= array
+    ]
+  toEncoding (Input type' name location array) = pairs
+    (  "type" .= type'
+    <> "name" .= name
+    <> "location" .= location
+    <> "array" .= array
+    )
+
+data Ubo = Ubo
+  { name :: Text
+  , set :: Int
+  , binding :: Int
+  , array :: Maybe (Vector Int)
+  } deriving (Show, Generic, FromJSON, ToJSON)
+
+data Texture = Texture
+  { type' :: TextureDescriptorType
+  , name :: Text
+  , set :: Int
+  , binding :: Int
+  , array :: Maybe (Vector Int)
+  } deriving (Show, Generic)
+
+instance FromJSON Texture where
+  parseJSON = withObject "textures" $ \v -> Texture
+    <$> v .: "type"
+    <*> v .: "name"
+    <*> v .: "set"
+    <*> v .: "binding"
+    <*> v .:? "array"
+
+instance ToJSON Texture where
+  toJSON (Texture type' name set binding array) = object
+    [ "type" .= type'
+    , "name" .= name
+    , "set" .= set
+    , "binding" .= binding
+    , "array" .= array
+    ]
+  toEncoding (Texture type' name set binding array) = pairs
+    (  "type" .= type'
+    <> "name" .= name
+    <> "set" .= set
+    <> "binding" .= binding
+    <> "array" .= array
+    )
+
+data Reflection = Reflection
+  { entryPoints :: Vector EntryPoint
+  , inputs :: Maybe (Vector Input)
+  , textures :: Maybe (Vector Texture)
+  , ubos :: Maybe (Vector Ubo)
+  } deriving (Show, Generic, FromJSON, ToJSON)
+
+data ConvertException = ConvertException Text Text
+  deriving (Show)
+
+instance Exception ConvertException
+
+class Convert a where
+  eitherFrom :: Text -> Either ConvertException a
+  from :: Text -> a
+  from x = case eitherFrom x of
```

Comment:
Embedding this in directly in a class isn't necessary.

It looks like you're extending "withText" with a mapping from a subset of text values to constructors. This really should be in Data.Aeson.Types...

```haskell
withTextMaybe :: String -> (Text -> Maybe a) -> Value -> Parser a
withTextMaybe label fromText =
  withText label $ \text ->
    case fromText text of
      Nothing ->
        parseFail $ "Unexpected " <> label <> " value: " <> show text
      Just known ->
        pure known
```

With that you can define instances without defining *any* new classes:

```haskell
instance FromJSON Example where
  parseJSON = withTextMaybe "Example" exampleFromText

exampleFromText :: Text -> Maybe Example
exampleFromText = \case ...

instance ToJSON Example where
   toJSON = toJSON . exampleToText

exampleToText :: Example -> Text
exampleToText = \case ...
```

Reply:
thanks for reviewing!

I fixed the issue in json parse relate code in the way your suggestion. But I am wondering why create a `Convert a` class is not recommend. 

For example, `Convert a` class can help prevent naming functions with long name like `vertexAttributeTypeFromText`

Reply:
Long names are fine. You type them once and then you can get a type, definition, usage instances etc.

Type classes add a level of indirection, trading off some of the benefits of standalone functions. This may be worth it when you expect your code to be extended elsewhere... But all the class-related code is contained in the same module.

And sometimes you even have to fight typechecker with annotations, defeating its "shorter names" purpose.